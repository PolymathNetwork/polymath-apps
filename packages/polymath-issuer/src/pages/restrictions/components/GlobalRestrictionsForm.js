// @flow

import React, { Component, useState, useEffect, Fragment } from 'react';
import { connect } from 'react-redux';
import { Modal, Button } from '@polymathnetwork/ui';
import { withFormik } from 'formik';
import { Form, Dropdown, DropdownItem } from 'carbon-components-react';
import { RESTRICTION_TYPE } from '../../../constants';
import {
  bull,
  Box,
  PageCentered,
  ContentBox,
  Heading,
  FormItem,
  TextInput,
  RadioInput,
  PercentageInput,
  NumberInput,
  Grid,
  FormItemGroup,
  DatePickerInput,
  TimePickerSelect,
} from '@polymathnetwork/ui';
import validator from '@polymathnetwork/ui/validator';
import {
  addDefaultRestriction,
  addDefaultDailyRestriction,
  modifyDefaultDailyRestriction,
  modifyDefaultRestriction,
} from '../../../actions/restrictions';
import {
  validateTodayOrAfter,
  validateStartTime,
  validateEndDate,
  validateEndTime,
  REQUIRED_MESSAGE,
  MORE_THAN_MESSAGE,
  ADDRESS_MESSAGE,
  MAX_DIGITS_MESSAGE,
} from '../../sto/components/ConfigureSTOForm/validators'; // this is mangled put into global object
import moment from 'moment';
import { toWei } from '../../../utils/contracts';

type Props = {
  isOpen: boolean,
  handleClose: () => any,
};

const formSchema = validator.object().shape({
  date: validator.object().shape({
    startDate: validator.date().isRequired(REQUIRED_MESSAGE),
    startTime: validator.number().isRequired(REQUIRED_MESSAGE),
    endDate: validator
      .date()
      .isRequired(REQUIRED_MESSAGE)
      .test('validateEndDate', validateEndDate),
    endTime: validator
      .number()
      .isRequired(REQUIRED_MESSAGE)
      .test('validEndTime', validateEndTime),
  }),
  token: validator.number().when('transferType', {
    is: 'token',
    then: validator.number().isRequired(REQUIRED_MESSAGE),
  }),
  percentage: validator.number().when('transferType', {
    is: 'percentage',
    then: validator.number().isRequired(REQUIRED_MESSAGE),
  }),
  intervalAmount: validator
    .number()
    .when('restrictionType', {
      is: 'custom',
      then: validator.number().isRequired(REQUIRED_MESSAGE),
    })
    .when('restrictionType', {
      is: '24h',
      then: validator.number().nullable(),
    }),
});

const initialValues = {
  date: {
    startDate: null,
    startTime: null,
    endDate: null,
    endTime: null,
  },
  intervalAmount: null,
  transferType: 'token',
  token: null,
  percentage: '',
  interval: 'days',
};

export const AddGlobalRestrictionsComponent = ({
  handleSubmit,
  handleClose,
  handleChange,
  values,
  setFieldValue,
  errors,
  touched,
}) => {
  const handleDropdown = e => {
    setFieldValue('interval', e.value);
  };
  return (
    <Form className="global-restrictions" onSubmit={handleSubmit}>
      <Grid>
        <Grid.Row>
          <Grid.Col gridSpan={5}>
            <FormItem name="transferType">
              <FormItem.Label>
                <strong>Specify Maximum Transfer in</strong>
              </FormItem.Label>
              <FormItem.Input
                className="transfer-type"
                options={[
                  { label: 'Number of Tokens', value: 'token' },
                  { label: 'Percentage', value: 'percentage' },
                ]}
                component={RadioInput}
              />
              <FormItem.Error />
            </FormItem>
          </Grid.Col>
          <Grid.Col gridSpan={7}>
            {values.transferType === 'token' && (
              <FormItem name="token">
                <Heading className="form-header" variant="h3">
                  Maximum Trade Volume
                </Heading>
                <FormItem.Input
                  placeholder="Enter the value"
                  component={NumberInput}
                  unit="TOKEN"
                />
                <FormItem.Error />
              </FormItem>
            )}
            {values.transferType === 'percentage' && (
              <FormItem name="percentage">
                <Heading className="form-header" variant="h3">
                  Maximum Trade Volume
                </Heading>
                <FormItem.Input
                  placeholder="Enter the value"
                  component={PercentageInput}
                  unit="%"
                />
                <FormItem.Error />
              </FormItem>
            )}
          </Grid.Col>
        </Grid.Row>
        {values.restrictionType === 'custom' && (
          <Fragment>
            <Grid.Row>
              <Grid.Col gridSpan={12}>
                <label className="form-label">Rolling Period Interval</label>
              </Grid.Col>
            </Grid.Row>
            <Grid.Row style={{ marginTop: '-20px' }}>
              <Grid.Col gridSpan={12}>
                <FormItemGroup>
                  <FormItemGroup.Items>
                    <FormItem name="intervalAmount">
                      <FormItem.Input
                        component={NumberInput}
                        placeholder="Enter amount"
                        maxDecimals={0}
                        min={1}
                        max={365}
                      />
                      <FormItem.Error />
                    </FormItem>
                    <FormItem name="interval">
                      <FormItem.Input
                        className="align-self-end"
                        component={Dropdown}
                        onChange={handleDropdown}
                        value={values.interval}
                        ariaLabel="Rolling Period Interval"
                      >
                        <DropdownItem value="days" itemText="Days" />
                        <DropdownItem value="months" itemText="Months" />
                        <DropdownItem value="years" itemText="Years" />
                      </FormItem.Input>
                    </FormItem>
                  </FormItemGroup.Items>
                </FormItemGroup>
              </Grid.Col>
              <Grid.Col className="align-self-end" gripSpan={2} />
            </Grid.Row>
          </Fragment>
        )}
        <Grid.Row>
          <Grid.Col gridSpan={12}>
            <FormItemGroup>
              <FormItemGroup.Items>
                <FormItem name="date.startDate">
                  <FormItem.Label>Start Date</FormItem.Label>
                  <FormItem.Input
                    component={DatePickerInput}
                    placeholder="mm / dd / yyyy"
                  />
                </FormItem>
                <FormItem name="date.startTime">
                  <FormItem.Label>Time</FormItem.Label>
                  <FormItem.Input
                    component={TimePickerSelect}
                    placeholder="hh:mm"
                  />
                </FormItem>
                <FormItem name="date.endDate">
                  <FormItem.Label>End Date</FormItem.Label>
                  <FormItem.Input
                    component={DatePickerInput}
                    placeholder="mm / dd / yyyy"
                  />
                </FormItem>
                <FormItem name="date.endTime">
                  <FormItem.Label>Time</FormItem.Label>
                  <FormItem.Input
                    component={TimePickerSelect}
                    placeholder="hh:mm"
                  />
                </FormItem>
              </FormItemGroup.Items>
              <FormItemGroup.Error
                name="date"
                errors={errors}
                touched={touched}
              />
            </FormItemGroup>
          </Grid.Col>
        </Grid.Row>
        <Grid.Row>
          <Grid.Col gridSpan={4} />
        </Grid.Row>
      </Grid>
      <Button type="submit">Set The Period</Button>
    </Form>
  );
};

const formikEnhancer = withFormik({
  validationSchema: formSchema,
  displayName: 'GlobalRestrictionsForm',
  validateOnChange: false,
  mapPropsToValues: props => {
    if (
      props.restrictionType === 'custom' &&
      props.defaultRestrictionModified
    ) {
      let startTime =
        props.defaultRestriction.startTime.unix() -
        moment(props.defaultRestriction.startTime)
          .startOf('day')
          .unix();
      let endTime =
        props.defaultRestriction.endTime.unix() -
        moment(props.defaultRestriction.endTime)
          .startOf('day')
          .unix();
      return {
        date: {
          startDate: props.defaultRestriction.startTime,
          startTime: startTime * 1000,
          endDate: props.defaultRestriction.endTime,
          endTime: endTime * 1000,
        },
        intervalAmount: props.defaultRestriction.rollingPeriodInDays,
        transferType:
          props.defaultRestriction.restrictionType === 1
            ? 'percentage'
            : 'token',
        token:
          props.defaultRestriction.restrictionType === 0
            ? parseFloat(props.defaultRestriction.allowedTokens)
            : '',
        percentage:
          props.defaultRestriction.restrictionType === 1
            ? parseFloat(props.defaultRestriction.allowedTokens)
            : '',
        interval: 'days',
        restrictionType: props.restrictionType,
      };
    } else if (
      props.restrictionType === '24h' &&
      props.dailyRestrictionModified
    ) {
      let startTime =
        props.dailyRestriction.startTime.unix() -
        moment(props.dailyRestriction.startTime)
          .startOf('day')
          .unix();
      let endTime =
        props.dailyRestriction.endTime.unix() -
        moment(props.dailyRestriction.endTime)
          .startOf('day')
          .unix();
      return {
        date: {
          startDate: props.dailyRestriction.startTime,
          startTime: startTime * 1000,
          endDate: props.dailyRestriction.endTime,
          endTime: endTime * 1000,
        },
        transferType:
          props.dailyRestriction.restrictionType === 1 ? 'percentage' : 'token',
        token:
          props.dailyRestriction.restrictionType === 0
            ? parseFloat(props.dailyRestriction.allowedTokens)
            : '',
        percentage:
          props.dailyRestriction.restrictionType === 1
            ? parseFloat(props.dailyRestriction.allowedTokens)
            : '',
        restrictionType: props.restrictionType,
      };
    }
    return {
      ...initialValues,
      restrictionType: props.restrictionType,
    };
  },
  handleSubmit: (values, { errors, setFieldError, props }) => {
    const {
      dispatch,
      handleClose,
      dailyRestrictionModified,
      defaultRestrictionModified,
    } = props;
    const startsAt =
      moment(values.date.startDate).unix() * 1000 + values.date.startTime;
    const endsAt =
      moment(values.date.endDate).unix() * 1000 + values.date.endTime;
    const allowedTokens =
      values.transferType === 'token'
        ? toWei(values.token)
        : toWei(values.percentage);
    let rollingPeriodInDays;

    switch (values.interval) {
      case 'months':
        rollingPeriodInDays = values.intervalAmount * 30;
        break;
      case 'years':
        rollingPeriodInDays = values.intervalAmount * 365;
        break;
      default:
        rollingPeriodInDays = values.intervalAmount;
        break;
    }

    handleClose();

    if (values.restrictionType === '24h') {
      if (!dailyRestrictionModified) {
        dispatch(
          addDefaultDailyRestriction(
            allowedTokens,
            startsAt,
            endsAt,
            RESTRICTION_TYPE[values.transferType]
          )
        );
      } else {
        dispatch(
          modifyDefaultDailyRestriction(
            allowedTokens,
            startsAt,
            endsAt,
            RESTRICTION_TYPE[values.transferType]
          )
        );
      }
    } else {
      if (!defaultRestrictionModified) {
        dispatch(
          addDefaultRestriction(
            allowedTokens,
            startsAt,
            rollingPeriodInDays,
            endsAt,
            RESTRICTION_TYPE[values.transferType]
          )
        );
      } else {
        dispatch(
          modifyDefaultRestriction(
            allowedTokens,
            startsAt,
            rollingPeriodInDays,
            endsAt,
            RESTRICTION_TYPE[values.transferType]
          )
        );
      }
    }
  },
});

const mapStateToProps = state => ({
  dailyRestrictionModified: state.restrictions.dailyRestrictionModified,
  defaultRestrictionModified: state.restrictions.defaultRestrictionModified,
  dailyRestriction: state.restrictions.dailyRestriction,
  defaultRestriction: state.restrictions.defaultRestriction,
});

const FormikEnhancedForm = formikEnhancer(AddGlobalRestrictionsComponent);
const ConnectedForm = connect(mapStateToProps)(FormikEnhancedForm);

export default ConnectedForm;
