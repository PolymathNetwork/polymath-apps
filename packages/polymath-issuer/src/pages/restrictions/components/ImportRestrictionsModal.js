import React, { Component } from 'react';
import { connect } from 'react-redux';
import {
  Icon,
  FileUploader,
  Button,
  InlineNotification,
} from 'carbon-components-react';
import { Modal, Remark, Paragraph } from '@polymathnetwork/ui';
import {
  uploadCSV,
  resetUploaded,
  addIndividualRestrictionMulti,
} from '../../../actions/restrictions';

import { fetch } from '../../../actions/sto';
import { STAGE_OVERVIEW } from '../../../reducers/sto';

import type { RootState } from '../../../redux/reducer';

type StateProps = {|
  isTooMany: boolean,
  parseError: String,
  isReady: boolean,
  isInvalid: boolean,
|};

type DispatchProps = {|
  fetch: () => any,
  addIndividualRestrictionMulti: () => any,
  uploadCSV: (file: Object) => any,
  resetUploaded: () => any,
|};

type Props = {|
  isOpen: boolean,
  onClose: () => any,
  onSubmit: (file: ?Object) => any,
|} & StateProps &
  DispatchProps;

const mapStateToProps = state => ({
  isTooMany: state.restrictions.isTooMany,
  parseError: state.restrictions.parseError,
  isReady: state.restrictions.uploaded.length > 0,
  isInvalid: state.restrictions.criticals.length > 0,
  sto: state.sto,
});

const mapDispatchToProps = {
  fetch,
  uploadCSV,
  resetUploaded,
  addIndividualRestrictionMulti,
};

class ImportRestrictionsModal extends Component<Props> {
  componentDidMount() {
    this.props.fetch();
  }

  handleClose = () => {
    this.fileUploader.clearFiles();
    this.props.resetUploaded();
    this.props.onClose();
  };

  handleSubmit = () => {
    this.handleClose();
    this.props.addIndividualRestrictionMulti();
  };

  handleUploaded = async (event: Object) => {
    const file = event.target.files[0];
    if (file.type.match(/csv.*/) || file.name.match(/.*\.csv$/i)) {
      await this.props.uploadCSV(file);
      //NOTE @sajclarke: This hack is necessary to add an eventlistener to the dynamic filename container from FileUploader
      const node = this.fileUploader.nodes[0];
      if (node) {
        const el = Array.from(node.getElementsByClassName('bx--file-close'))[0];
        el.addEventListener(
          'click',
          e => {
            this.props.resetUploaded();
          },
          false
        );
      }
    }
  };

  fileUploaderRef = (el: ?Object) => {
    // $FlowFixMe
    this.fileUploader = el;
  };

  render() {
    const {
      isOpen,
      sto,
      isTooMany,
      parseError,
      isReady,
      isInvalid,
    } = this.props;

    return (
      <Modal
        isOpen={isOpen}
        onClose={this.handleClose}
        className="whitelist-import-modal"
      >
        <Modal.Header>Import Whitelist</Modal.Header>
        <h4 className="pui-h4">
          Add multiple addresses to the whitelist by uploading a comma separated
          .CSV file.{' '}
          <strong>You may add up to 40 addresses per .CSV file</strong>. The
          format should be as follows:
          <br />• ETH Address (address to volume restrict);
          <br />• 24h Start Date: <strong>mm/dd/yyyy</strong> (date when the
          volume restrictions will start for that address);
          <br />• 24h End Date: <strong> mm/dd/yyyy</strong> (date when the
          volume restrictions will end for that address);
          <br />• 24h Restriction Type: <strong>0 or 1</strong>;<br />• Whether
          the allowed tokens will be a percentage of all tokens or a fix value.
          Set to <strong>1</strong> for percentage.
          <br />• 24h Allowed Tokens: <strong>0 to 100</strong> when{' '}
          <strong>Restriction Type is 1</strong>
          <br />
          <br />
        </h4>
        <h5 className="pui-h5">
          You can&nbsp;&nbsp;&nbsp;
          <Icon name="download" fill="#252D6B" width="16" height="16" />
          &nbsp;
          <a href="/restrictions-sample.csv" className="pui-bold" download>
            Download Sample.csv
          </a>
          &nbsp;&nbsp;file and edit it.
        </h5>
        <FileUploader
          iconDescription="Cancel"
          multiple={false}
          buttonLabel="Upload File"
          onChange={this.handleUploaded}
          className="file-uploader"
          accept={['.csv']}
          buttonKind="secondary"
          filenameStatus="edit"
          ref={this.fileUploaderRef}
        />
        {parseError.length > 0 ? (
          <InlineNotification
            hideCloseButton
            title={parseError}
            subtitle="Please check instructions above and try again."
            kind="error"
          />
        ) : isInvalid ? (
          <InlineNotification
            hideCloseButton
            title="The file you uploaded has invalid fields"
            subtitle="Please check instructions above and try again."
            kind="error"
          />
        ) : isInvalid && !isReady ? (
          <InlineNotification
            hideCloseButton
            title="The file you uploaded does not contain any valid values"
            subtitle="Please check instructions above and try again."
            kind="error"
          />
        ) : isTooMany ? (
          <InlineNotification
            hideCloseButton
            title="The file you uploaded contains more than 40 investors"
            subtitle="You can still continue, but only 40 first investors will be submitted."
            kind="error"
          />
        ) : (
          <Remark title="Reminder">
            Investors must be approved before they are added to the whitelist.{' '}
            <br />
            Your file cannot exceed 40 addresses. If you have more than 40
            addresses on your whitelist, upload multiple files.
          </Remark>
        )}
        <Modal.Footer>
          <Paragraph align="right">
            <Button
              className="cancel-btn"
              kind="secondary"
              onClick={this.handleClose}
            >
              Cancel
            </Button>
            <Button
              type="submit"
              kind="primary"
              disabled={!isReady || isInvalid}
              onClick={this.handleSubmit}
            >
              Import Whitelist
            </Button>
          </Paragraph>
        </Modal.Footer>
      </Modal>
    );
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ImportRestrictionsModal);
