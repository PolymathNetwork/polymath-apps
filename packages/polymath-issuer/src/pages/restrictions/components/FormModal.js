// @flow

import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withFormik } from 'formik';
import { Form } from 'carbon-components-react';
import { Remark, Modal, Button } from '@polymathnetwork/ui';
import validator from '@polymathnetwork/ui/validator';
import { addAddressToTransferManager } from '../../../actions/compliance';
import GlobalRestrictionsForm from './GlobalRestrictionsForm';

type Props = {
  isOpen: boolean,
  handleClose: () => any,
  restrictionType: string,
};

export class FormModal extends Component<Props> {
  render() {
    const { isOpen, handleClose, restrictionType } = this.props;
    return (
      <Modal isOpen={isOpen} onClose={handleClose}>
        <Modal.Body>
          {restrictionType === 'custom' && (
            <h1 className="pui-h2">
              Configure Custom Trade Volume Restriction
            </h1>
          )}
          {restrictionType === '24h' && (
            <h1 className="pui-h2">Configure Daily Trade Volume Restriction</h1>
          )}
          <h3 className="pui-h3">
            The volume restriction can be specified in number of tokens or as a
            percentage of total supply and will be enforced only between the
            Start date and time and the End date and time.
          </h3>
          <Remark title="Note">
            — Volume restrictions do not apply to primary issuance, only to
            secondary trades.
            <br />— The Global Restriction applies to each Investor
            individually, not the aggregate of all sales across all investors
            within the period.
          </Remark>
          <GlobalRestrictionsForm
            handleClose={handleClose}
            restrictionType={restrictionType}
          />
        </Modal.Body>
      </Modal>
    );
  }
}

export default FormModal;
