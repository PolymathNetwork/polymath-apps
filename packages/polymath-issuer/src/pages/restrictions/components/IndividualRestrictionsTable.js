// @flow

import React, { Component } from 'react';
import { DataTable, Icon } from 'carbon-components-react';
import { Button } from '@polymathnetwork/ui';
import IndividualRestrictionsModal from './IndividualRestrictionsModal';
import ImportRestrictionsModal from './ImportRestrictionsModal';
import { connect } from 'react-redux';
import { modifyIndividualRestriction } from '../../../actions/restrictions';
import { dispatch } from 'rxjs/internal/observable/range';
import moment from 'moment';
const {
  TableContainer,
  Table,
  TableHead,
  TableRow,
  TableBody,
  TableCell,
  TableHeader,
  TableToolbar,
  TableToolbarContent,
} = DataTable;

const columns = [
  {
    header: 'Investor Wallet Address',
    key: 'address',
    Cell: ({ value }) => value,
  },
  {
    header: 'Maximum 24h Transfer (# or %)',
    key: 'dailyRestrictionType',
    Cell: ({ value }) => value,
  },
  {
    header: 'Start Date',
    key: 'dailyStartTime',
    cell: ({ value }) => value,
  },
  {
    header: 'End Date',
    key: 'dailyEndTime',
    Cell: ({ value }) => value,
  },
  {
    header: 'Custom Rolling Period Interval',
    key: 'rollingPeriodInDays',
    Cell: ({ value }) => value,
  },
  {
    header: 'Maximum Custom Transfer (# or %)',
    key: 'customRestrictionType',
    Cell: ({ value }) => value,
  },
  {
    header: 'Start Date',
    key: 'customStartTime',
    Cell: ({ value }) => value,
  },
  {
    header: 'End Date',
    key: 'customEndTime',
    Cell: ({ value }) => value,
  },
];

type State = {|
  isIndividualRestrictionModalOpen: boolean,
  isUploadCsvModal: boolean,
|};

class IndividualRestrictionsTable extends Component<Props, State> {
  state = {
    isIndividualRestrictionModalOpen: false,
    isUploadCsvModal: false,
  };

  editAddress = address => {
    const { individualRestrictions } = this.props;
    let restriction = individualRestrictions.find(i => i.address === address);
    this.props.modifyIndividualRestriction(restriction);
    this.setState({ isIndividualRestrictionModalOpen: true });
  };

  handleCsvOpen = () => {
    this.setState({ isUploadCsvModal: true });
  };

  handleCsvClose = () => {
    this.setState({ isUploadCsvModal: false });
  };

  handleOpen = () => {
    this.setState({ isIndividualRestrictionModalOpen: true });
  };

  handleClose = () => {
    this.setState({ isIndividualRestrictionModalOpen: false });
    this.props.modifyIndividualRestriction(null);
  };

  handleDelete = id => {
    this.props.removeAddressFromTransferManager(id);
  };

  formatCell = (id, cell) => {
    const { individualRestrictions } = this.props;
    if (cell.value === undefined) return '-';
    switch (cell.info.header) {
      case 'customStartTime':
      case 'dailyStartTime':
      case 'customEndTime':
      case 'dailyEndTime':
        return moment.unix(cell.value / 1000).format('MMM DD YYYY');
      case 'customRestrictionType':
        let restriction = individualRestrictions.find(i => i.address === id);
        return `${
          cell.value == 1
            ? restriction.customAllowedTokens * 100
            : restriction.customAllowedTokens
        } ${cell.value == 1 ? '%' : 'TOKENS'}`;
      case 'dailyRestrictionType':
        let x = individualRestrictions.find(i => i.address === id);
        return `${
          cell.value == 1 ? x.dailyAllowedTokens * 100 : x.dailyAllowedTokens
        } ${cell.value == 1 ? '%' : 'TOKENS'}`;
      case 'rollingPeriodInDays':
        return `${cell.value} days`;
      default:
        return cell.value;
    }
  };

  render() {
    const { individualRestrictions } = this.props;
    return (
      <div>
        <IndividualRestrictionsModal
          isOpen={this.state.isIndividualRestrictionModalOpen}
          handleClose={this.handleClose}
        />
        <ImportRestrictionsModal
          isOpen={this.state.isUploadCsvModal}
          onClose={this.handleCsvClose}
        />
        <DataTable
          headers={columns}
          rows={
            individualRestrictions < 1
              ? [{ id: '0', address: '-', details: '-' }]
              : individualRestrictions
          }
          render={({ rows, headers, getHeaderProps }) => {
            return (
              <TableContainer title="">
                <TableToolbar>
                  <TableToolbarContent>
                    {/* pass in `onInputChange` change here to make filtering work */}
                    <Button onClick={this.handleOpen} icon="icon--add">
                      Add new
                    </Button>
                    <Button
                      style={{ marginLeft: '5px' }}
                      onClick={this.handleCsvOpen}
                      icon="upload"
                    >
                      Upload CSV
                    </Button>
                  </TableToolbarContent>
                </TableToolbar>
                <Table>
                  <TableHead>
                    <TableRow>
                      {headers.map(header => (
                        <TableHeader
                          key={header}
                          {...getHeaderProps({ header })}
                        >
                          {header.header}
                        </TableHeader>
                      ))}
                    </TableRow>
                  </TableHead>
                  <TableBody>
                    {rows.map(row => (
                      <TableRow
                        key={row.id}
                        onClick={() => this.editAddress(row.id)}
                      >
                        {row.cells.map(cell => (
                          <TableCell key={cell.id}>
                            {this.formatCell(row.id, cell)}
                          </TableCell>
                        ))}
                      </TableRow>
                    ))}
                  </TableBody>
                </Table>
              </TableContainer>
            );
          }}
        />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  individualRestrictions: state.restrictions.individualRestrictions,
});

const mapDispatchToProps = {
  modifyIndividualRestriction,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(IndividualRestrictionsTable);
