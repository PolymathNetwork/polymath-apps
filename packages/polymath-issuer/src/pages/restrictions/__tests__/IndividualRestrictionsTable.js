import { IndividualRestrictionsTable } from '../components/IndividualRestrictionsTable';
import renderer from 'react-test-renderer';
import React from 'react';
import configureStore from 'redux-mock-store';
import { theme, GlobalStyles } from '@polymathnetwork/ui';
import { shallow, mount, render } from 'enzyme';
import { ThemeProvider } from 'styled-components';
import moment from 'moment';
import { Provider } from 'react-redux';
const mockStore = configureStore([]);

describe('Individual Restrictions Table Component', () => {
  it('should render correctly', () => {
    const props = {
      theme,
      individualRestrictions: [
        {
          address: '0x53d284357ec70cE289D6D64134DfAc8E511c8a3D',
          id: '0x53d284357ec70cE289D6D64134DfAc8E511c8a3D',
          customStartTime: '1641013200000',
          customEndTime: '1717473600000',
          rollingPeriodInDays: '20',
          customRestrictionType: '0',
          customAllowedTokens: '120',
          dailyStartTime: '1641013200000',
          dailyEndTime: '1717473600000',
          dailyRestrictionType: '1',
          dailyAllowedTokens: '1',
        },
        {
          address: '0x66f820a414680B5bcda5eECA5dea238543F42054',
          id: '0x66f820a414680B5bcda5eECA5dea238543F42054',
          customStartTime: '1641013200000',
          customEndTime: '1717473600000',
          rollingPeriodInDays: '10',
          customRestrictionType: '0',
          customAllowedTokens: '1000',
        },
        {
          address: '0xd9f346Bf88cA2cb7e11B0106018DE80A0169764D',
          id: '0xd9f346Bf88cA2cb7e11B0106018DE80A0169764D',
          customStartTime: '1641013200',
          customEndTime: '1717473600',
          rollingPeriodInDays: '10',
          customRestrictionType: '0',
          customAllowedTokens: '1000',
        },
        {
          address: '0xE2FACfbdEdB03EeBC6747498aea18f7bc97A7066',
          id: '0xE2FACfbdEdB03EeBC6747498aea18f7bc97A7066',
          dailyStartTime: '1569902400',
          dailyEndTime: '1570077000',
          dailyRestrictionType: '0',
          dailyAllowedTokens: '100',
        },
        {
          address: '0xCaA749f1cf0110dbfCA59e3594EB845a166DE232',
          id: '0xCaA749f1cf0110dbfCA59e3594EB845a166DE232',
          dailyStartTime: '1569558600',
          dailyEndTime: '1569821400',
          dailyRestrictionType: '0',
          dailyAllowedTokens: '100',
        },
        {
          address: '0x1B2D2797A56c9eBa12aa821D643aB0dBa66C3772',
          id: '0x1B2D2797A56c9eBa12aa821D643aB0dBa66C3772',
          customStartTime: '1641013200',
          customEndTime: '1717473600',
          rollingPeriodInDays: '20',
          customRestrictionType: '0',
          customAllowedTokens: '120',
          dailyStartTime: '1641013200',
          dailyEndTime: '1717473600',
          dailyRestrictionType: '0',
          dailyAllowedTokens: '1000',
        },
        {
          address: '0xC5fdf4076b8F3A5357c5E395ab970B5B54098Fef',
          id: '0xC5fdf4076b8F3A5357c5E395ab970B5B54098Fef',
          dailyStartTime: '1569902400',
          dailyEndTime: '1570078800',
          dailyRestrictionType: '0',
          dailyAllowedTokens: '100',
        },
        {
          address: '0xE853c56864A2ebe4576a807D26Fdc4A0adA51919',
          id: '0xE853c56864A2ebe4576a807D26Fdc4A0adA51919',
          dailyStartTime: '1569994200',
          dailyEndTime: '1572494400',
          dailyRestrictionType: '0',
          dailyAllowedTokens: '1200',
        },
        {
          address: '0x3f517cF8cf3E7b8149Db333b8f3393Cb57bA1d3B',
          id: '0x3f517cF8cf3E7b8149Db333b8f3393Cb57bA1d3B',
          dailyStartTime: '1569988800',
          dailyEndTime: '1572409800',
          dailyRestrictionType: '0',
          dailyAllowedTokens: '100',
        },
      ],
    };
    const tree = shallow(<IndividualRestrictionsTable {...props} />);
    expect(tree).toMatchSnapshot();
  });
});
