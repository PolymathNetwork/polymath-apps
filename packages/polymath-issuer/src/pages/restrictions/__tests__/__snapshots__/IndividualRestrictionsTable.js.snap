// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Individual Restrictions Table Component should render correctly 1`] = `
<div>
  <IndividualRestrictionsModal
    handleClose={[Function]}
    isOpen={false}
  />
  <Connect(ImportRestrictionsModal)
    isOpen={false}
    onClose={[Function]}
  />
  <DataTable
    filterRows={[Function]}
    headers={
      Array [
        Object {
          "Cell": [Function],
          "header": "Investor Wallet Address",
          "key": "address",
        },
        Object {
          "Cell": [Function],
          "header": "Maximum 24h Transfer (# or %)",
          "key": "dailyRestrictionType",
        },
        Object {
          "cell": [Function],
          "header": "Start Date",
          "key": "dailyStartTime",
        },
        Object {
          "Cell": [Function],
          "header": "End Date",
          "key": "dailyEndTime",
        },
        Object {
          "Cell": [Function],
          "header": "Custom Rolling Period Interval",
          "key": "rollingPeriodInDays",
        },
        Object {
          "Cell": [Function],
          "header": "Maximum Custom Transfer (# or %)",
          "key": "customRestrictionType",
        },
        Object {
          "Cell": [Function],
          "header": "Start Date",
          "key": "customStartTime",
        },
        Object {
          "Cell": [Function],
          "header": "End Date",
          "key": "customEndTime",
        },
      ]
    }
    locale="en"
    render={[Function]}
    rows={
      Array [
        Object {
          "address": "0x53d284357ec70cE289D6D64134DfAc8E511c8a3D",
          "customAllowedTokens": "120",
          "customEndTime": "1717473600000",
          "customRestrictionType": "0",
          "customStartTime": "1641013200000",
          "dailyAllowedTokens": "1",
          "dailyEndTime": "1717473600000",
          "dailyRestrictionType": "1",
          "dailyStartTime": "1641013200000",
          "id": "0x53d284357ec70cE289D6D64134DfAc8E511c8a3D",
          "rollingPeriodInDays": "20",
        },
        Object {
          "address": "0x66f820a414680B5bcda5eECA5dea238543F42054",
          "customAllowedTokens": "1000",
          "customEndTime": "1717473600000",
          "customRestrictionType": "0",
          "customStartTime": "1641013200000",
          "id": "0x66f820a414680B5bcda5eECA5dea238543F42054",
          "rollingPeriodInDays": "10",
        },
        Object {
          "address": "0xd9f346Bf88cA2cb7e11B0106018DE80A0169764D",
          "customAllowedTokens": "1000",
          "customEndTime": "1717473600",
          "customRestrictionType": "0",
          "customStartTime": "1641013200",
          "id": "0xd9f346Bf88cA2cb7e11B0106018DE80A0169764D",
          "rollingPeriodInDays": "10",
        },
        Object {
          "address": "0xE2FACfbdEdB03EeBC6747498aea18f7bc97A7066",
          "dailyAllowedTokens": "100",
          "dailyEndTime": "1570077000",
          "dailyRestrictionType": "0",
          "dailyStartTime": "1569902400",
          "id": "0xE2FACfbdEdB03EeBC6747498aea18f7bc97A7066",
        },
        Object {
          "address": "0xCaA749f1cf0110dbfCA59e3594EB845a166DE232",
          "dailyAllowedTokens": "100",
          "dailyEndTime": "1569821400",
          "dailyRestrictionType": "0",
          "dailyStartTime": "1569558600",
          "id": "0xCaA749f1cf0110dbfCA59e3594EB845a166DE232",
        },
        Object {
          "address": "0x1B2D2797A56c9eBa12aa821D643aB0dBa66C3772",
          "customAllowedTokens": "120",
          "customEndTime": "1717473600",
          "customRestrictionType": "0",
          "customStartTime": "1641013200",
          "dailyAllowedTokens": "1000",
          "dailyEndTime": "1717473600",
          "dailyRestrictionType": "0",
          "dailyStartTime": "1641013200",
          "id": "0x1B2D2797A56c9eBa12aa821D643aB0dBa66C3772",
          "rollingPeriodInDays": "20",
        },
        Object {
          "address": "0xC5fdf4076b8F3A5357c5E395ab970B5B54098Fef",
          "dailyAllowedTokens": "100",
          "dailyEndTime": "1570078800",
          "dailyRestrictionType": "0",
          "dailyStartTime": "1569902400",
          "id": "0xC5fdf4076b8F3A5357c5E395ab970B5B54098Fef",
        },
        Object {
          "address": "0xE853c56864A2ebe4576a807D26Fdc4A0adA51919",
          "dailyAllowedTokens": "1200",
          "dailyEndTime": "1572494400",
          "dailyRestrictionType": "0",
          "dailyStartTime": "1569994200",
          "id": "0xE853c56864A2ebe4576a807D26Fdc4A0adA51919",
        },
        Object {
          "address": "0x3f517cF8cf3E7b8149Db333b8f3393Cb57bA1d3B",
          "dailyAllowedTokens": "100",
          "dailyEndTime": "1572409800",
          "dailyRestrictionType": "0",
          "dailyStartTime": "1569988800",
          "id": "0x3f517cF8cf3E7b8149Db333b8f3393Cb57bA1d3B",
        },
      ]
    }
    sortRow={[Function]}
    translateWithId={[Function]}
  />
</div>
`;
